saleor/discount/migrations/tasks/saleor3_16.py

```diff
@@ -278,28 +292,54 @@ def migrate_translations(sale_ids, saleid_promotion_map):
 
 
 def migrate_checkout_line_discounts(sale_ids, rule_by_channel_and_sale):
-    if checkout_line_discounts := CheckoutLineDiscount.objects.filter(
+    checkout_line_discounts = CheckoutLineDiscount.objects.filter(
         sale_id__in=sale_ids
-    ).select_related("line__checkout"):
-        for checkout_line_discount in checkout_line_discounts:
+    ).exclude(type="promotion")[:CHECKOUT_LINE_DISCOUNT_BATCH_SIZE]
+    discount_line_ids = list(checkout_line_discounts.values_list("id", flat=True))
+    if discount_line_ids:
+        line_discounts = CheckoutLineDiscount.objects.filter(
+            id__in=discount_line_ids
+        ).select_related("line__checkout")
+        discounts_to_update = []
+        for checkout_line_discount in line_discounts:
+            discounts_to_update.append(checkout_line_discount)
+            checkout_line_discount.type = "promotion"
             if checkout_line := checkout_line_discount.line:
                 channel_id = checkout_line.checkout.channel_id
                 sale_id = checkout_line_discount.sale_id
                 lookup = f"{channel_id}_{sale_id}"
-                checkout_line_discount.type = "promotion"
                 if promotion_rule := rule_by_channel_and_sale.get(lookup):
                     checkout_line_discount.promotion_rule = promotion_rule
 
         CheckoutLineDiscount.objects.bulk_update(
-            checkout_line_discounts, ["promotion_rule_id", "type"]
+            discounts_to_update, ["promotion_rule_id", "type"]
         )
+        migrate_checkout_line_discounts(sale_ids, rule_by_channel_and_sale)
```

Comment:
I believe we need to change recursion into loop as we may hit recursion limit.

Reply:
I changed it to celery task